src/main.c

```diff
@@ -30,19 +30,20 @@
 
 #include "org.freedesktop.login1.Session.h"
 #include "org.freedesktop.login1.h"
+#include "sdi-helpers.h"
 #include "sdi-refresh-monitor.h"
+#include "sdi-snapd-monitor.h"
 #include "sdi-theme-monitor.h"
 
 static Login1Manager *login_manager = NULL;
 static SnapdClient *client = NULL;
 static SdiThemeMonitor *theme_monitor = NULL;
 static SdiRefreshMonitor *refresh_monitor = NULL;
 static gboolean running = FALSE;
+static SdiSnapdMonitor *snapd_monitor = NULL;
 
-static gchar *snapd_socket_path = NULL;
-
-static GOptionEntry entries[] = {{"snapd-socket-path", 0, 0,
-                                  G_OPTION_ARG_FILENAME, &snapd_socket_path,
+static GOptionEntry entries[] = {{"snapd-socket-path", 0, G_OPTION_FLAG_NONE,
+                                  G_OPTION_ARG_FILENAME, &sdi_snapd_socket_path,
```

Comment:
@robert-ancell I need to have access to this parameter from the `helpers.c` file, to be able to create as many snapd-clients as needed using the right path. Which way do you think is better to do it?

* this way, exporting the `gchar *` variable from the `helpers.c` file and using it directly here
* keeping that string private to the `helpers.c` file and use a function to copy that parameter into the private variable, calling it in `do_startup()`
* moving the whole GOptionEntry variable into `helpers.c` and add a function, also in `helpers.c` that receives the GApplication variable and calls `g_application_add_main_option_entries(G_APPLICATION(app), entries);`

Reply:
`helpers.c` shouldn't have any state, so moving the variable into it suggests this is not good architecture. In general, having helpers indicates that these functions aren't in the right place - they should be owned by a class or module with clear purpose.

It seems to me there should be a `sdi_application_create_snapd_client` and a `SdiApplication` class that owns `snapd_socket_path`.

Reply:
Uf... creating a gobject singleton just for that is a little bit... I think that I'll create a "factory module".

Reply:
Thanks!